Enable dynamic SHM support

lorc opened the pull request:
Hello.

There are last patches for dynamic SHM. 

I'm not sure, if I did right thing, when created two patches with ABI changes, because 4e5d00d ("ABI change: add shared memory support") enables new capability and 30260ae ("ABI change: enable TMEM parameters with non-contiguous shm") extends it.

Probably I should squash those two patches into one, or create one additional patch for `entry_fast.c`, that enables `OPTEE_SMC_SEC_CAP_DYNAMIC_SHM`. What do you think?

stuyoder commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -290,6 +290,12 @@
 #define OPTEE_SMC_SEC_CAP_HAVE_RESERVED_SHM	(1 << 0)
 /* Secure world can communicate via previously unregistered shared memory */
 #define OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM	(1 << 1)
+/*
```
Based on the mailing list discussion, we don't need both OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM and OPTEE_SMC_SEC_CAP_DYNAMIC_SHM, right?

Jens had proposed changing the existing "UNREGISTERED_SHM" capability to this:

/*
  * Secure world supports shared memory objects to represent
  * physically non-contiguous memory and the commands
  * "register/unregister shared memory object" for more efficient
  * communication via physically non-contiguous memory.
  */
#define OPTEE_SMC_SEC_CAP_NONCONTIG_SHM  (1 << 1)

lorc replied:
Hi, @stuyoder. I did exactly this. I decided not to touch existing `OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM`. Instead I added new `OPTEE_SMC_SEC_CAP_DYNAMIC_SHM`. I think, that this name better describes this capability.

etienne-lms replied:
Actually the UNREGISTERED_SHM relates the the so-called default shm (`NSEC_SHM` in the core). This shm is "negotiated" at core inits with nsec world (some SMCs allow to get/set the nsec_shm location, but core_mmu.c expects it from static config `CFG_SHMEM_START/_SIZE`). It happens to be physically contiguous and as a single memory chunk. That's its real attributes.

stuyoder replied:
What do you mean with respect to "negotiated"?  UNREGISTERED_SHM is unused as of now in both OP-TEE and Linux.

This is how the capabilities are defined today:

/* Secure world has reserved shared memory for normal world to use */
#define OPTEE_SMC_SEC_CAP_HAVE_RESERVED_SHM     (1 << 0)
/* Secure world can communicate via previously unregistered shared memory */
#define OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM      (1 << 1)

It is HAVE_RESERVED_SHM that is referring to the physically contiguous static CFG_SHMEM_START/SIZE.

The question is what capabilities we need to refer to the new ability for OP-TEE to map arbitrary, non-contiguous physical addresses.

Do we need:
A).  One capability that refers to the capability of OP-TEE to handle non-contiguous shared memory and the capability to handle the register/unregister core APIs?
or
B) . Two capabilities-- 1) one for the capability of OP-TEE to handle non-contiguous shared memory, and 2) one for the capability to handle the register/unregister core APIs.

Jens proposal on the mailing list was for option #A.

stuyoder replied:
@lorc, what do the capabilities OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM and OPTEE_SMC_SEC_CAP_DYNAMIC_SHM mean in your view?  How are they different?

As I stated above, the first question is whether we need one capability or two capabilities.  Once that is decided, we can decide what the best name is.

jenswi-linaro replied:
I'd prefer only on capability as those two are so intertwined.

lorc replied:
@stuyoder, I'm not using `OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM` at all. I have leaved it as is. I use only `OPTEE_SMC_SEC_CAP_DYNAMIC_SHM` in my patch series.
I decided not to rename `OPTEE_SMC_SEC_CAP_UNREGISTERED_SHM` to `OPTEE_SMC_SEC_CAP_DYNAMIC_SHM` because that can add some confusion. 
At this moment there are at least 30 free bits for capability flags, so I don't see why we need to reuse existing values.

prime-zeng commented:
@lorc Do you have any plan to support user space mapping for the dynamic SHM?

jenswi-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -94,8 +114,26 @@ static TEE_Result assign_mobj_to_param_mem(const paddr_t pa, const size_t sz,
 }
 
+static TEE_Result set_rmem_param(const struct optee_msg_param *param,
+			  struct param_mem *mem)
```
Align with `(` on the line above.

jenswi-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -94,8 +114,26 @@ static TEE_Result assign_mobj_to_param_mem(const paddr_t pa, const size_t sz,
+
 static TEE_Result copy_in_params(const struct optee_msg_param *params,
-		uint32_t num_params, struct tee_ta_param *ta_param)
+			 uint32_t num_params, struct tee_ta_param *ta_param,
```
Align with `(` on the line above.

jenswi-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -342,6 +422,80 @@ static void entry_cancel(struct thread_smc_args *smc_args,
+			mobj_free(mobj);
+			arg->ret = TEE_SUCCESS;
+		} else {
+			EMSG("Can't find mapping with given cookie\n");
```
`\n` not needed.

lorc commented:
Addressed comments.

@prime-zeng, I'm sorry. I didn't get what are you asking. 
User TAs can use memory that was registered by CA, if this is what you asked.

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -76,10 +76,30 @@ static bool param_mem_from_mobj(struct param_mem *mem, struct mobj *mobj,
+		return TEE_SUCCESS;
+	}
+
+	/* Non-contigous buffer from non sec DDR? */
```
apology: could you add a space before the `?` for consistency with below. (which could be rephrased some day).

jforissier replied:
No space before `?` in English...

etienne-lms replied:
So we'll clean the other comments.

lorc replied:
I added separate patch that does this.

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -76,10 +76,30 @@ static bool param_mem_from_mobj(struct param_mem *mem, struct mobj *mobj,
+							  false);
+		if (!mem->mobj)
+			return TEE_ERROR_BAD_PARAMETERS;
+		mem->offs = (pa & SMALL_PAGE_MASK);
```
minor: useless parenthesis.

etienne-lms commented:
> > @lorc Do you have any plan to support user TA mapping for the dynamic SHM?
>
> @prime-zeng, I'm sorry. I didn't get what are you asking.
> User TAs can use memory that was registered by CA, if this is what you asked.

I agree with @lorc. Since his work on the shm, the memref parameters of TA invocations are now handled through memory object `mobj`. These abstract to underlying memory when one operate on the memory (identification, mapping, address conversion,  ...). The current implementation of the mapping of user TA memref parameters should be able to map the new "mobj_reg_shm` reference (registered, not physically contiguous) as well as currently used `mobj_shm` (from the NSEC_SHM). Yet, let's see the tests, but it should get functional.

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -108,14 +147,12 @@ static TEE_Result copy_in_params(const struct optee_msg_param *params,
+		if (saved_attr[n] & OPTEE_MSG_ATTR_META)
 			return TEE_ERROR_BAD_PARAMETERS;
-		if (params[n].attr & OPTEE_MSG_ATTR_NONCONTIG)
-			return TEE_ERROR_BAD_PARAMETERS;
```
I think we should reject noncontig attributes.
Non-contig reference are used only from `register_[un]shm()` std entries to register a reference related to the parameter type `OPTEE_MSG_ATTR_RMEM_*`.
Current function `copy_params()` is called from `open/close_session()` and and `invoke_command()`. It does not expect non-contig references, only value (with.. value), tmem (with its phys addr) and rmem (with its registered ref) types.

jenswi-linaro replied:
You mean we should disable noncontig bufs as parameters to TA unless it's via already registered buffers? Why?
I think it could be useful for one-shot invokes etc.

etienne-lms replied:
> I think it could be useful for one-shot invokes etc.

You are right. It could be used.
So it would be allowed only for `OPTEE_MSG_ATTR_TMEM_*` types, as used in `assign_mobj_to_param_mem()`.

But TEE Client API expects client to either use `TEEC_AllocateSharedMemory()` or `TEE_RegisterSharedMemory(),` do we really need to support here un-registered shm references ?

jenswi-linaro replied:
There's the `TEEC_MEMREF_TEMP_*` also.

prime-zeng commented:
@etienne-lms @lorc I mean to ask whether the User TAs can use the registered (not physically contiguous )memory. And i see the final User TAs mapping are created in the function `set_pg_region` ,but i think this functions can't deal with non-contiguous memory.  Do i miss something?

etienne-lms commented:
> `set_pg_region()` ,but i think this functions can't deal with non-contiguous memory. Do i miss something?

As of current implementation, yes, the function `set_pg_region()` may not fully handle the non-contiguous references. But it is already based on a loop that maps page per page and relies on `mobj_get_pa()` to retrieve each page physical address. So I think it is a matter of debugging to insure non-contiguous memory references are properly handled, if not already supported.

prime-zeng commented:
@etienne-lms   i got it, thank you, so i think maybe the problem now is how to fix the  ` set_pg_region() `  to support non-contiguous memory, and this is another issue.

lorc commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -379,7 +489,17 @@ void __weak tee_entry_std(struct thread_smc_args *smc_args)
+			return;
+		}
+		mobj = map_cmd_buffer(parg, &num_params);
+	}
```
I prefer to leave it in current state. `get_cmd_buffer()` is responsible for buffers in `MEM_AREA_NSEC_SHM` which is already mapped in OP-TEE address space. `map_cmd_buffer()` maps buffers from `MEM_AREA_RAM_NSEC`. One task for one function.

As alternative I can create a third function that will choose to call  `get_cmd_buffer()` or `map_cmd_buffer()`.

lorc commented:
@prime-zeng , actually, one of the changes makes `set_pg_region()`  (at [1]) to  support non-contiguous memory. I especially tested this case.

@etienne-lms, I have addressed your comments.

[1] https://github.com/OP-TEE/optee_os/pull/1631/files#diff-21c5a372f38c092a429fd2b7e8f71ef3R1060

jenswi-linaro commented:
I'd like to run some tests with this, which driver and client branch did you use?
Also we need to be happy about especially the driver changes before we commit to this new ABI.

lorc commented:
Hi @jenswi-linaro, you will need kernel patches from this PR: https://github.com/linaro-swg/linux/pull/29
And client library patches from this PR: https://github.com/OP-TEE/optee_client/pull/72

jenswi-linaro commented:
This works very well. The ABI provided to normal world is flexible enough to handle all use cases and I'm pretty sure it will survive a review on the kernel mailing lists without incompatible changes. 

I'd like to wrap this up and merge it as soon as possible. We should do the same for  linaro-swg/linux#29 and OP-TEE/optee_client#72. When it comes to upstreaming the kernel changes I'd be grateful I you posted the patches @lorc, in order to drive the process on the kernel mailing lists.

etienne-lms commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -103,6 +103,13 @@ static void tee_entry_exchange_capabilities(struct thread_smc_args *args)
 	args->a1 = OPTEE_SMC_SEC_CAP_HAVE_RESERVED_SHM;
+
+	if (core_mmu_nsec_ddr_is_defined()) {
+		IMSG("NS DDR configured. Enabling dynamic SHM");
```
minor: s/configured/defined/. This code does not really configure anything.
Same below.

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -76,16 +76,36 @@ static bool param_mem_from_mobj(struct param_mem *mem, struct mobj *mobj,
+		return TEE_SUCCESS;
+	}
+
+	/* belongs to nonsecure shared memory? */
```
minor: s/belongs/Belongs/

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -76,16 +76,36 @@ static bool param_mem_from_mobj(struct param_mem *mem, struct mobj *mobj,
 
 #ifdef CFG_SECURE_DATA_PATH
-	/* belongs to SDP memories ? */
+	/* belongs to SDP memories? */
```
minor: s/belongs/Belongs/

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
@lorc could you please squash/rebase the patches and apply the Reviewed-by: tags?

Can this safely be merged without the linux and optee_client patches? I'm concerned by the "ABI change:" text. Assuming the changes are compatible, I'd suggest changing the subject of two commits as follows:
- "ABI change: add shared memory support" would become "core: add support for dynamic registration of shared memory"
- "ABI change: enable TMEM parameters with non-contiguous shm" would become "core: allow non-contiguous shared memory in temporary memory references"

Did I get this right?

lorc commented:
Hello.

I have squashed, rebased and applied tags to this patch series.
But there is one new patch, that was born during kernel part rework: `core: enable TMEM parameters with non-contiguous shm` (https://github.com/OP-TEE/optee_os/pull/1631/commits/eb3e932d705b58d727ed198613f360f6e582dbec). I didn't applied tags to it, because it wasn't reviewed. I don't know what to do with it. Will you review it as a part of this series, or should I post it as a separate PR?

jenswi-linaro commented:
You can add:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`
to "core: enable TMEM parameters with non-contiguous shm"

lorc commented:
Thank you, @jenswi-linaro , I added your R-b tag and repushed patches.

jforissier commented:
Sorry to bother, but I'm still not happy with the commit subject: "core: add shared memory support". OP-TEE already supports shared memory so it doesn't make sense. Say waht you are really adding.

Same goes with "core: enable TMEM parameters with non-contiguous shm". What is TMEM? This acronym appears absolutely nowhere, not even in that patch.

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -292,6 +325,9 @@ static void entry_open_session(struct thread_smc_args *smc_args,
 	plat_prng_add_jitter_entropy();
 
 out:
+	cleanup_params(arg->params + num_meta, saved_attr,
```
Can you cross check that there is no issue when this cleanup from `get_open_session_meta()`. `num_meta` may not be inited ?

stuyoder commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -346,6 +385,77 @@ static void entry_cancel(struct thread_smc_args *smc_args,
+		return NULL;
+	}
+
+	*num_params = arg->num_params;
```
FYI...note there is no barrier (i.e. dsb) between the call to mobj_mapped_shm_alloc() and where the arg->num_params variable is read.  The missing dsb() causes a data abort on my system.  This is the issue raised in https://github.com/OP-TEE/optee_os/issues/1805.

jenswi-linaro replied:
Thanks for the reminder, I've created https://github.com/OP-TEE/optee_os/pull/1827 to deal with it.

etienne-lms commented:
@lorc, When testing the latest dynamic shm patches on my HW I faced issues that is related to 32bit mmu (non-lpae). Please have a look at https://github.com/OP-TEE/optee_os/pull/1831, issue is in the current implementation of optee_os, not in your patches.

Default qemu_virt setup do test non-lpae mapping and it successfully tests the dynamic shm setup. I would have expect it to fail for the same reason that mandated https://github.com/OP-TEE/optee_os/pull/1831 on my hw.

lorc commented:
@etienne-lms, thank you for testing. 

I tested it only on qemu v7, qemu v8 and on RCAR H3 (wich is also ARMv8). It is really strange that qemu v7 passed this test, while real HW is not. But I'm glad that you have found this issue. And thank you for the fix.

etienne-lms commented:
Without some memory access firewalling, the issue raises on cache content due to the secure tag of line. Qemu does not handle cache, nor memory firewalling upon NS state, so the issue may not be detected.

stuyoder commented:
@etienne-lms do you have the patch in https://github.com/OP-TEE/optee_os/pull/1827 applied?

For some reason the issue only manifested itself when running in 32 bit mode on real hardware.  It may have nothing to do with your issue, but thought I would check.  You have to have #1827 applied.

etienne-lms commented:
@stuyoder, yes I tested with the memory barrier patch applied (actually I did not even try to test without :)

jforissier commented:
@lorc thanks for re-wording the commit subjects.

Can people provide some Tested-by's? I'll give it a try on HiKey and/or HiKey960.

jenswi-linaro commented:
It doesn't work properly on FVP due to the kernel using physical addresses above 4G. I have fixes for AArch64, looking into AArch32 now.

jforissier commented:
Doesn't work on HiKey either :( Boot hangs.
Git bisect points at c6c3b59 ("core: add registered shared memory support").

lorc commented:
Hmm, that's interesting. Does it hangs during optee-os boot or during linux kernel boot?

jforissier commented:
Kernel.
Hangs there:
```
[    0.688034] uart-pl011 f7112000.uart: could not find pctldev for node /soc/pinmux@f7010000/uart2_pmx_func, deferring probe
```

Successful boot shows:
```
[    0.680020] uart-pl011 f7112000.uart: could not find pctldev for node /soc/pinmux@f7010000/uart2_pmx_func, deferring probe
[    1.440310] f7113000.uart: ttyAMA3 at MMIO 0xf7113000 (irq = 10, base_baud = 0) is a PL011 rev2
[    1.449157] console [ttyAMA3] enabled
```
...but it's hard to tell where it hangs exactly because in any case the console does not show as much as `dmesg`, here is what `dmesg` shows with optee_os master (relevant part only):
```
...
[    0.680020] uart-pl011 f7112000.uart: could not find pctldev for node /soc/pinmux@f7010000/uart2_pmx_func, deferring probe
[    0.691374] uart-pl011 f7113000.uart: could not find pctldev for node /soc/pinmux@f7010000/uart3_pmx_func, deferring probe
[    0.696251] hi6220-mbox f7510000.mailbox: Mailbox enabled
[    0.735708] HugeTLB registered 2 MB page size, pre-allocated 0 pages
[    0.739598] ACPI: Interpreter disabled.
[    0.740360] vgaarb: loaded
[    0.740620] SCSI subsystem initialized
[    0.740820] libata version 3.00 loaded.
[    0.740925] ssp-pl022 f7106000.spi: could not find pctldev for node /soc/pinmux@f7010000/spi0_pmx_func, deferring probe
[    0.741281] usbcore: registered new interface driver usbfs
[    0.741390] usbcore: registered new interface driver hub
[    0.741532] usbcore: registered new device driver usb
[    0.741735] i2c_designware f7100000.i2c: could not find pctldev for node /soc/pinmux@f7010000/i2c0_pmx_func, deferring probe
[    0.741883] i2c_designware f7101000.i2c: could not find pctldev for node /soc/pinmux@f7010000/i2c1_pmx_func, deferring probe
[    0.742027] i2c_designware f7102000.i2c: could not find pctldev for node /soc/pinmux@f7010000/i2c2_pmx_func, deferring probe
[    0.742339] pps_core: LinuxPPS API ver. 1 registered
[    0.742401] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    0.742523] PTP clock support registered
[    0.742694] dmi: Firmware registration failed.
[    0.742826] Registered efivars operations
[    0.742995] Advanced Linux Sound Architecture Driver Initialized.
[    0.743841] clocksource: Switched to clocksource arch_sys_counter
[    0.785350] VFS: Disk quotas dquot_6.6.0
[    0.785462] VFS: Dquot-cache hash table entries: 512 (order 0, 4096 bytes)
[    0.785774] pnp: PnP ACPI: disabled
[    0.792089] NET: Registered protocol family 2
[    0.792613] TCP established hash table entries: 8192 (order: 4, 65536 bytes)
[    0.792754] TCP bind hash table entries: 8192 (order: 5, 131072 bytes)
[    0.792943] TCP: Hash tables configured (established 8192 bind 8192)
[    0.793066] UDP hash table entries: 512 (order: 2, 16384 bytes)
[    0.793156] UDP-Lite hash table entries: 512 (order: 2, 16384 bytes)
[    0.793356] NET: Registered protocol family 1
[    0.793720] RPC: Registered named UNIX socket transport module.
[    0.793793] RPC: Registered udp transport module.
[    0.793851] RPC: Registered tcp transport module.
[    0.793909] RPC: Registered tcp NFSv4.1 backchannel transport module.
[    0.793989] PCI: CLS 0 bytes, default 128
[    0.794182] Unpacking initramfs...
[    1.278375] Freeing initrd memory: 11864K (ffff800036f0f000 - ffff800037aa5000)
[    1.278984] kvm [1]: 8-bit VMID
[    1.279032] kvm [1]: IDMAP page: 9e2000
[    1.279081] kvm [1]: HYP VA range: 800000000000:ffffffffffff
[    1.279960] kvm [1]: Hyp mode initialized successfully
[    1.280062] kvm [1]: vgic-v2@f6804000
[    1.280311] kvm [1]: vgic interrupt IRQ1
[    1.280392] kvm [1]: virtual timer IRQ4
[    1.282535] futex hash table entries: 2048 (order: 6, 262144 bytes)
[    1.282717] audit: initializing netlink subsys (disabled)
[    1.282839] audit: type=2000 audit(1.143:1): initialized
[    1.283338] workingset: timestamp_bits=46 max_order=18 bucket_order=0
[    1.290591] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    1.291273] NFS: Registering the id_resolver key type
[    1.291356] Key type id_resolver registered
[    1.291410] Key type id_legacy registered
[    1.291467] nfs4filelayout_init: NFSv4 File Layout Driver Registering...
[    1.291696] 9p: Installing v9fs 9p2000 file system support
[    1.293810] Block layer SCSI generic (bsg) driver version 0.4 loaded (major 246)
[    1.293907] io scheduler noop registered
[    1.294060] io scheduler cfq registered (default)
[    1.295000] libphy: mdio_driver_register: phy-bcm-ns2-pci
[    1.296093] pinctrl-single f7010000.pinmux: 159 pins at pa ffff000009139000 size 636
[    1.296562] pinctrl-single f7010800.pinmux: 163 pins at pa ffff00000913b800 size 652
[    1.296790] pinctrl-single f8001800.pinmux: 30 pins at pa ffff00000913d800 size 120
[    1.297908] pl061_gpio f8011000.gpio: PL061 GPIO chip @0x00000000f8011000 registered
[    1.298386] pl061_gpio f8012000.gpio: PL061 GPIO chip @0x00000000f8012000 registered
[    1.298862] pl061_gpio f8013000.gpio: PL061 GPIO chip @0x00000000f8013000 registered
[    1.299009] gpio gpiochip3: gpio-line-names specifies 9 line names but there are 8 lines on the chip
[    1.299466] pl061_gpio f8014000.gpio: PL061 GPIO chip @0x00000000f8014000 registered
[    1.299949] pl061_gpio f7020000.gpio: PL061 GPIO chip @0x00000000f7020000 registered
[    1.300423] pl061_gpio f7021000.gpio: PL061 GPIO chip @0x00000000f7021000 registered
[    1.300888] pl061_gpio f7022000.gpio: PL061 GPIO chip @0x00000000f7022000 registered
[    1.301345] pl061_gpio f7023000.gpio: PL061 GPIO chip @0x00000000f7023000 registered
[    1.301502] gpio gpiochip8: gpio-line-names specifies 9 line names but there are 8 lines on the chip
[    1.301937] pl061_gpio f7024000.gpio: PL061 GPIO chip @0x00000000f7024000 registered
[    1.302403] pl061_gpio f7025000.gpio: PL061 GPIO chip @0x00000000f7025000 registered
[    1.302875] pl061_gpio f7026000.gpio: PL061 GPIO chip @0x00000000f7026000 registered
[    1.303345] pl061_gpio f7027000.gpio: PL061 GPIO chip @0x00000000f7027000 registered
[    1.303824] pl061_gpio f7028000.gpio: PL061 GPIO chip @0x00000000f7028000 registered
[    1.304314] pl061_gpio f7029000.gpio: PL061 GPIO chip @0x00000000f7029000 registered
[    1.304784] pl061_gpio f702a000.gpio: PL061 GPIO chip @0x00000000f702a000 registered
[    1.305257] pl061_gpio f702b000.gpio: PL061 GPIO chip @0x00000000f702b000 registered
[    1.305724] pl061_gpio f702c000.gpio: PL061 GPIO chip @0x00000000f702c000 registered
[    1.306179] pl061_gpio f702d000.gpio: PL061 GPIO chip @0x00000000f702d000 registered
[    1.306635] pl061_gpio f702e000.gpio: PL061 GPIO chip @0x00000000f702e000 registered
[    1.307102] pl061_gpio f702f000.gpio: PL061 GPIO chip @0x00000000f702f000 registered
[    1.311459] xenfs: not registering filesystem on non-xen platform
[    1.315993] Serial: 8250/16550 driver, 4 ports, IRQ sharing disabled
[    1.317731] SuperH (H)SCI(F) driver initialized
[    1.318121] msm_serial: driver initialized
[    1.325494] loop: module loaded
[    1.328595] hisi_sas: driver version v1.6
[    1.330233] libphy: Fixed MDIO Bus: probed
[    1.330574] tun: Universal TUN/TAP device driver, 1.6
[    1.330638] tun: (C) 1999-2004 Max Krasnyansky <maxk@qualcomm.com>
[    1.331324] e1000e: Intel(R) PRO/1000 Network Driver - 3.2.6-k
[    1.331395] e1000e: Copyright(c) 1999 - 2015 Intel Corporation.
[    1.331508] igb: Intel(R) Gigabit Ethernet Network Driver - version 5.4.0-k
[    1.331589] igb: Copyright (c) 2007-2014 Intel Corporation.
[    1.331695] igbvf: Intel(R) Gigabit Virtual Function Network Driver - version 2.4.0-k
[    1.331791] igbvf: Copyright (c) 2009 - 2012 Intel Corporation.
[    1.331925] sky2: driver version 1.30
[    1.332224] usbcore: registered new interface driver asix
[    1.332336] usbcore: registered new interface driver ax88179_178a
[    1.332437] usbcore: registered new interface driver cdc_ether
[    1.332535] usbcore: registered new interface driver dm9601
[    1.332631] usbcore: registered new interface driver net1080
[    1.332729] usbcore: registered new interface driver cdc_subset
[    1.332828] usbcore: registered new interface driver zaurus
[    1.332937] usbcore: registered new interface driver cdc_ncm
[    1.333143] VFIO - User Level meta-driver version: 0.3
[    1.335185] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    1.335286] ehci-pci: EHCI PCI platform driver
[    1.335378] ehci-platform: EHCI generic platform driver
[    1.335514] ehci-exynos: EHCI EXYNOS driver
[    1.335623] ehci-msm: Qualcomm On-Chip EHCI Host Controller
[    1.335742] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    1.335872] ohci-pci: OHCI PCI platform driver
[    1.335970] ohci-platform: OHCI generic platform driver
[    1.336092] ohci-exynos: OHCI EXYNOS driver
[    1.336432] usbcore: registered new interface driver usb-storage
[    1.337236] mousedev: PS/2 mouse device common for all mice
[    1.337717] input: HISI 65xx PowerOn Key as /devices/platform/f8000000.pmic/hi65xx-powerkey.0.auto/input/input0
[    1.338454] rtc-pl031 f8003000.rtc: rtc core: registered pl031 as rtc0
[    1.338752] rtc-pl031 f8004000.rtc: rtc core: registered pl031 as rtc1
[    1.339165] i2c /dev entries driver
[    1.341645] sdhci: Secure Digital Host Controller Interface driver
[    1.341730] sdhci: Copyright(c) Pierre Ossman
[    1.341961] Synopsys Designware Multimedia Card Interface Driver
[    1.342492] dwmmc_k3 f723d000.dwmmc0: fifo-depth property not found, using value of FIFOTH register as default
[    1.342720] dwmmc_k3 f723d000.dwmmc0: IDMAC supports 32-bit address mode.
[    1.343125] dwmmc_k3 f723d000.dwmmc0: Using internal DMA controller.
[    1.343208] dwmmc_k3 f723d000.dwmmc0: Version ID is 250a
[    1.343310] dwmmc_k3 f723d000.dwmmc0: DW MMC controller at irq 40,32 bit host data width,256 deep fifo
[    1.359896] mmc_host mmc0: Bus speed (slot 0) = 24800000Hz (slot req 400000Hz, actual 400000HZ div = 31)
[    1.379912] dwmmc_k3 f723d000.dwmmc0: 1 slots initialized
[    1.380270] dwmmc_k3 f723e000.dwmmc1: fifo-depth property not found, using value of FIFOTH register as default
[    1.380512] dwmmc_k3 f723e000.dwmmc1: IDMAC supports 32-bit address mode.
[    1.380683] dwmmc_k3 f723e000.dwmmc1: Using internal DMA controller.
[    1.380764] dwmmc_k3 f723e000.dwmmc1: Version ID is 250a
[    1.380869] dwmmc_k3 f723e000.dwmmc1: DW MMC controller at irq 41,32 bit host data width,128 deep fifo
[    1.381169] dwmmc_k3 f723e000.dwmmc1: Got CD GPIO
[    1.404026] mmc_host mmc1: Bus speed (slot 0) = 24800000Hz (slot req 400000Hz, actual 400000HZ div = 31)
[    1.424732] dwmmc_k3 f723e000.dwmmc1: 1 slots initialized
[    1.426263] dwmmc_k3 f723f000.dwmmc2: fifo-depth property not found, using value of FIFOTH register as default
[    1.426482] dwmmc_k3 f723f000.dwmmc2: IDMAC supports 32-bit address mode.
[    1.426680] dwmmc_k3 f723f000.dwmmc2: Using internal DMA controller.
[    1.426761] dwmmc_k3 f723f000.dwmmc2: Version ID is 250a
[    1.426855] dwmmc_k3 f723f000.dwmmc2: DW MMC controller at irq 42,32 bit host data width,128 deep fifo
[    1.427340] sdhci-pltfm: SDHCI platform and OF driver helper
[    1.429397] ledtrig-cpu: registered to indicate activity on CPUs
[    1.430207] usbcore: registered new interface driver usbhid
[    1.430285] usbhid: USB HID core driver
[    1.433291] mmc_host mmc0: Bus speed (slot 0) = 51756522Hz (slot req 52000000Hz, actual 51756522HZ div = 0)
[    1.433566] optee: probing for conduit method from DT.
[    1.433903] optee: initialized driver
[    1.434871] NET: Registered protocol family 17
[    1.434997] 9pnet: Installing 9P2000 support
[    1.435102] Key type dns_resolver registered
[    1.435746] registered taskstats version 1
[    1.436026] mmc0: new high speed MMC card at address 0001
[    1.436888] mmcblk0: mmc0:0001 MMC04G 3.60 GiB 
[    1.437389] mmcblk0boot0: mmc0:0001 MMC04G partition 1 2.00 MiB
[    1.437768] mmcblk0boot1: mmc0:0001 MMC04G partition 2 2.00 MiB
[    1.438012] mmcblk0rpmb: mmc0:0001 MMC04G partition 3 128 KiB
[    1.439825] f7112000.uart: ttyAMA2 at MMIO 0xf7112000 (irq = 9, base_baud = 0) is a PL011 rev2
[    1.440310] f7113000.uart: ttyAMA3 at MMIO 0xf7113000 (irq = 10, base_baud = 0) is a PL011 rev2
[    1.449157] console [ttyAMA3] enabled
...
```
The console log level looks OK (first 7 means all messages should go to the console, right?):
```
root@HiKey:/ cat /proc/sys/kernel/printk
7	4	1	7
```
Maybe a consequence of the UART probing code.

Anyway, could very well hang in the optee driver init.

lorc commented:
I think it hangs during optee driver initialization, but you can't see this because UART driver is still not initialized. Only after successful UART driver initialization it spews out whole log buffer. 

I suspect that it hangs in `tee_entry_exchange_capabilities()`, or more precise in `core_mmu_nsec_ddr_is_defined()`. 

Can you see OP-TEE debug output?

jenswi-linaro commented:
https://github.com/OP-TEE/optee_os/pull/1860 takes care of the FVP problems, both for AArch64 and AArch32. A nice side effect is that we can map 64-bit physical addresses on AArch32.

jforissier commented:
@lorc found the problem. It's the traces in `tee_entry_exchange_capabilities()`. With the following patch, `xtest` passes on HiKey:

```diff
diff --git a/core/arch/arm/tee/entry_fast.c b/core/arch/arm/tee/entry_fast.c
index 24ce3d7..e85eef4 100644
--- a/core/arch/arm/tee/entry_fast.c
+++ b/core/arch/arm/tee/entry_fast.c
@@ -104,12 +104,8 @@ static void tee_entry_exchange_capabilities(struct thread_smc_args *args)
        args->a0 = OPTEE_SMC_RETURN_OK;
        args->a1 = OPTEE_SMC_SEC_CAP_HAVE_RESERVED_SHM;
 
-       if (core_mmu_nsec_ddr_is_defined()) {
-               IMSG("NS DDR defined. Enabling dynamic SHM");
+       if (core_mmu_nsec_ddr_is_defined())
                args->a1 |= OPTEE_SMC_SEC_CAP_DYNAMIC_SHM;
-       } else {
-               IMSG("No NS DDR defined for the platform. Disabling dynamic SHM");
-       }
 }
 
 static void tee_entry_disable_shm_cache(struct thread_smc_args *args)
```

I'm thinking the stack size might be too small, I need to check that.

lorc commented:
@jforissier,

did you tried to put `IMSG("some text\n")` to this function? Is it possible, that UART driver on HiKey tries to issue blocking calls or something like that, what is not appropriate in fast SMC call?

jforissier commented:
@lorc yep, `IMSG("")' as first line of 'tee_entry_exchange_capabilities()' hangs the boot. Funny thing is, the Linux kernel doesn't seem completely dead, because the heartbeat LED flashes. But user mode never gets initialized anyway.

lorc commented:
@jforissier, aha. So looks like it is caused by UART driver for HiKey. 
I don't know - is this right behavior? I can remove `IMSG()` calls in my patch, no problem. But the same problem might fire later in another place.

jforissier commented:
I have entered https://github.com/OP-TEE/optee_os/issues/1863 to deal with this problem.

My impression is that "something" (the kernel, probably) has disabled the UART or its clock during a small time window. That's between 0.68 and 1.44 in my message above (https://github.com/OP-TEE/optee_os/pull/1631#issuecomment-334531978). The optee driver init happens just inside this window, unfortunately. As a result, OP-TEE's `pl011_flush()` hangs, because it never obtains the "FIFO empty" flag.

I'm considering making the PL011 driver more robust to this kind of issue (which should really be non-fatal, this is only console debugging!) by limiting the loop count in `pl011_flush()`. I will open a PR.

In the mean time I think you can keep the `IMSG()` in your patch.

jforissier commented:
For "core: enable non-contiguous temporary reference parameters":

`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`

jforissier commented:
@lorc would you please:
- Introduce `CFG_DYN_SHM` (or `CFG_DYNAMIC_SHM` as you prefer), as suggested in https://github.com/OP-TEE/optee_os/pull/1866#discussion_r143563788? I propose it should be set to `?= y` by default, thus allowing platforms to use dynamic SHM simply by adding the proper `register_nsec_ddr()` line, and still allowing the feature to be disabled at compile time if need be.
- Rebase the patches and squash the fixup patch for a final review?

@jenswi-linaro @etienne-lms @stuyoder please monitor this PR and provide your Tested-by's / Acked-by's / Reviewed-by's when possible. Of course I'll make sure dependent PRs are merged before merging this one.

Thanks!

jenswi-linaro commented:
@jforissier, would you mind listing all the dependent PRs to make sure we're testing the right thing?

jforissier commented:
@jenswi-linaro this has to be tested with current optee_client branch master and current linaro-swg/linux branch optee (well, unless the platform needs a specific kernel), plus the following PRs:

- https://github.com/OP-TEE/optee_os/pull/1860 (required for FVP)
- https://github.com/OP-TEE/optee_os/pull/1866 and https://github.com/OP-TEE/optee_os/pull/1864 (required for Hikey)

Plus the proper `register_nsec_ddr()` of course. I think that's it, actually.

lorc commented:
@jforissier, also we will need this PR: https://github.com/linaro-swg/linux/pull/53 to ensure that variable length return buffers behave correctly. Without this patch GP tests will fail.

jforissier commented:
@lorc correct, I've just merged it, so no additional work needed. Thanks.

jenswi-linaro commented:
I've just pushed it to https://github.com/jenswi-linaro/linux-1/tree/optee-next also

jenswi-linaro commented:
```
Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (FVP, QEMU v7/v8)
```

lorc commented:
@jenswi-linaro, @jforissier  thank you for testing.
@jforissier, I squashed, rebased and added tags. Also I added patch that can disable `OPTEE_SMC_SEC_CAP_DYNAMIC_SHM`. I thing it is fine for debug purposes.

My first intention was to fully disable dynamic SHM. But that would require many careful changes in `entry_std.c` to fully support both configurations (+ new round of code review). I certainly can do this, but I'm not sure if it is needed.

Please let me know if you prefer dynamic SHM to be fully disabled.

etienne-lms commented:
> @jenswi-linaro @etienne-lms @stuyoder please monitor this PR and provide ...

Hi. I am testing the dynshm on my b2260 (32bit armv7).
For info find several issues with the later optee and the dynshm patches (I rebased on top of my optee_os). Some failures are not seems related to the dynshm (seems rather related to stack unwind) and I had to disable debug support. But a specific failure on a GP test (@jforissier  gp_9067 core make a page fault) disappears when I disabling the dynshm.

Actually, the issue seems systematic with dyn-shm + pager both enabled, not reproduced otherwise. I will try on the qemus with pager.

> Please let me know if you prefer dynamic SHM to be fully disabled.

On my side, I feel ok with the feature being disable when platform doest not register nsec physical area(s).

lorc commented:
> On my side, I feel ok with the feature being disable when platform doest not register nsec physical area(s).

Yeah, I thought about this. But, it does not eliminate unused code and also, strictly speaking, defined nsec region is not equal to enabled SHM. Yes, defined nsec memory region is prerequisite for dynamic SHM, but It can have other uses...

jforissier commented:
@lorc I think your `CFG_DBG_DISABLE_DYN_SHM_CAP` approach is good enough. At least, it fit my testing needs and is not too intrusive.

@etienne-lms relying on `register_nsec_ddr()` is not strictly correct as said above, and is also not convenient because it cannot be controlled from the command line (test environment...).
So I think we should stick to the simple `CFG_` approach in this PR, which has been discussed quite a lot already! And do further changes in a separate PR if needed.

That being said... for simplicity, I would recommend `CFG_DYN_SHM_CAP ?= y` and `#if defined(CFG_DYN_SHM_CAP)`. That is, reverse the logic and drop the useless `_DBG`.
Note also: s/nevertheles/nevertheless/

With that, commit "build: disable..." is:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
Actually, the last change (introduces CFG_...) does not remove dynamic shm support from the optee kernel, it only drop the capability flag. Hence even disabled, the REE can still register and use dyn-shm. It is the expected behaviour?

lorc commented:
Yep. I have pointed this in comments to `config.mk`:

```
 +# Note that you can disable this feature for debug purposes. OP-TEE will not
 +# report to Normal World that it support dynamic SHM. But, nevertheles it
 +# will accept dynamic SHM buffers.
```

etienne-lms commented:
:) i should have read further. Well, so this is the expected behaviour.
Thanks.

jforissier commented:
@etienne-lms yes and it's enough to exercise the usual shm pool on a platform that otherwise supports the dyn shm. `make CFG_DYN_SHM_CAP=n`, run `xtest` and you're done.

lorc commented:
@jforissier I did as you asked. That was trivial change, so I did it in place, if you don't mind. Also I added your `A-b`

jforissier commented:
@lorc thanks but you forgot to update the commit subject ;-)

lorc commented:
@jforissier oops :) Should be fine now. I also elaborated commit message a bit.

lorc commented:
`Tested-by: Volodymyr Babchuk <vlad.babchuk@gmail.com> (Rcar M3)` with #1871 
and with `CFG_UNWIND=n` (see #1872)

jforissier commented:
Shippable fails due to checkpatch warnings that can easily be fixed.

jenswi-linaro commented:
Together with https://github.com/OP-TEE/optee_os/pull/1874
`Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (Juno with and without pager)`

lorc commented:
@jforissier, should I fix this one?
```
WARNING: line over 80 characters
#95: FILE: core/arch/arm/tee/entry_fast.c:112:
+		IMSG("No NS DDR defined for the platform. Disabling dynamic SHM");
```
Kernel coding style recommends leave long text constants as is even if they are longer than 80 characters to ease up grepping:

> However, never break user-visible strings such as
> printk messages, because that breaks the ability to grep for them.

jforissier commented:
What about this (we want the message even when CFG_DYN_SHM_CAP is disabled):

```C
        bool dshm_en = false;

#ifdef CFG_DYN_SHM_CAP
        dhsm_en = core_mmu_nsec_ddr_is_defined();
#endif
        if (dhsm_en)
                args->a1 |= OPTEE_SMC_SEC_CAP_DYNAMIC_SHM;
        IMSG("Dynamic shared memory is %sabled", dhsm_en ? "en" : "dis");
```

lorc commented:
Okay, will do in this way.

lorc commented:
Reworked `tee_entry_exchange_capabilities()` and also added new `T-b` tags.

lorc commented:
Wow, I can't believe that it is done! Thank you to all of you guys for your support and help. I learned much while working on this feature. 
BTW, I already tested it with Xen and it allowed me to call OP-TEE from Dom0 without dirty hacks in the hypervisor. I'm going to upstream needed patches to Xen, so Dom0<->OP-TEE interaction will be a standard feature.
Next big task is to add virtualization support in OP-TEE: tag sessions, shared memory, loaded TAs with virtual machine ID, check accesses and so on.

etienne-lms commented:
congratulations @lorc (and all contribs) for this. took a while, thanks for your patience.
